enhancements/storage/csi-efs-operator.md

```diff
@@ -0,0 +1,335 @@
+
+### Upgrade from the community operator
+
+Prerequisites:
+* The community aws-efs-operator is installed in OCP 4.7 cluster in `openshift-operators` namespace. The CSI driver runs
+in the same namespace too.
+  * The operator `Subscription` contains the community catalog as the operator source.
+
+Expected workflow (confirmed with OLM team):
+
+1. User upgrades to 4.8. Nothing interesting happens at this point, because the operator `Subscription` still points
+   to the community source. The "old" operator is still running.
+2. User edits the `Subscription` and updates the `source:` to the Red Hat certified catalog.
+   (Alternatively, user may un-install the community operator and install the supported one.)
+3. OLM reconciles this change, checks the new catalog and if there is a newer operator there, it will install it as an
+   update. In other words, we need to ship a newer version in the Red Hat catalog than in the community one.
+4. The "new" CSI operator starts. It does not see `ClusterCSIDriver` CR for EFS and only emits Info level alert that
+   there is no CR and does nothing. This means the old CSI driver is still running (installed by the "old" operator)
+   and any apps that use it are still running.
+5. User reads release notes, documentation or alert and creates th CR.
+6. The "new" operator "adopts" the CSI driver objects. Namely, it updates DaemonSet of the CSI driver, which then does
+   rolling update of the driver pods, using supported images.
+
+To adopt the old objects, the new operator must use the same names for all objects created by the old operator.
+
+### [efs-utils](https://github.com/aws/efs-utils)
+
+We need to get efs-utils into the container with the EFS CSI driver to enable data encryption in transport. efs-utils
+are just two Python scripts, which create / monitor stunnel connections between OCP nodes and NFS server somewhere in
+AWS cloud.
+
+We've chosen to use base image approach. There are two reasons: 1) we know the drill, and 2) we don't want
+to support a generic RPM that can be used outside of OCP.
+
+1. Fork `github.com/aws/efs-utils` into `github.com/openshift/aws-efs-utils`.
+2. Create a new base image with the utilities, say `ose-aws-efs-utils-base`:
+   ```Dockerfile
+   FROM: ocp4.8:base
+   RUN yum install stunnel python <and any other deps>
+   COPY <the utilities> /usr/bin
+   ```
+3. The EFS CSI driver then uses it as the base image:
+   ```Dockerfile
+   FROM registry.ci.openshift.org/openshift/release:golang-1.15 AS builder
+   ... build the image ...
+
+   FROM ose-aws-efs-utils-base
+   COPY --from=builder <the driver binary> /usr/bin/
+   ```
+
+It would be nice if we did not ship `ose-aws-efs-utils-base` anywhere, so people can consume the EFS utils only through
+the EFS operator. **We do not want to maintain it for everyone.**
+
+### Code organization
+
+* Reuse existing `github.com/openshift/aws-efs-operator` for the operator.
```

Comment:
nit: add `repository` to not confuse with code

Reply:
IMO we should rename it to `aws-efs-csi-driver-operator` to be consistent with the other repositories.

Consistent naming is the only requirement for creating new CSI driver/operator repositories (we'd get away with that because this repository already exists, but it'd be nice to make it consistent).

Reply:
If it was a new repo, then aws-efs-csi-driver-operator would be great, now that it already exists, I think it's better to leave it as it is - who knows where is it referenced.

Reply:
Actually, a fork from aws-efs-operator to aws-efs-csi-driver-operator could make a lot of sense. That way the old community operator doesn't need to undergo what's essentially a full rewrite, or deal with the branching issue; and the new operator can have its preferred name but still have the benefit of preserving whatever code may be useful.

Reply:
Having looked at the [library-go port WIP](https://github.com/openshift/aws-efs-operator/pull/31), which has nothing in common with the original, I don't see a reason to try to reuse that repository.